datadog_checks_base/datadog_checks/checks/prometheus/mixins.py

```diff
@@ -345,134 +413,127 @@ def _extract_metric_from_map(self, _m, messages, obj_map, obj_help):
-        if message.name in self.label_joins:
-            matching_label = self.label_joins[message.name]['label_to_match']
+        if message.name in scraper_config['label_joins']:
+            matching_label = scraper_config['label_joins'][message.name]['label_to_match']
             for metric in message.metric:
                 labels_list = []
                 matching_value = None
                 for label in metric.label:
                     if label.name == matching_label:
                         matching_value = label.value
-                    elif label.name in self.label_joins[message.name]['labels_to_get']:
+                    elif label.name in scraper_config['label_joins'][message.name]['labels_to_get']:
                         labels_list.append((label.name, label.value))
                 try:
-                    self._label_mapping[matching_label][matching_value] = labels_list
+                    scraper_config['_label_mapping'][matching_label][matching_value] = labels_list
                 except KeyError:
                     if matching_value is not None:
-                        self._label_mapping[matching_label] = {matching_value: labels_list}
+                        scraper_config['_label_mapping'][matching_label] = {matching_value: labels_list}
 
-    def join_labels(self, message):
+    def _join_labels(self, message, scraper_config):
         # Filter metric to see if we can enrich with joined labels
-        if self.label_joins:
+        if scraper_config['label_joins']:
             for metric in message.metric:
                 for label in metric.label:
-                    if label.name in self._watched_labels:
+                    if label.name in scraper_config['_watched_labels']:
                         # Set this label value as active
-                        if label.name not in self._active_label_mapping:
-                            self._active_label_mapping[label.name] = {}
-                        self._active_label_mapping[label.name][label.value] = True
+                        if label.name not in scraper_config['_active_label_mapping']:
+                            scraper_config['_active_label_mapping'][label.name] = {}
+                        scraper_config['_active_label_mapping'][label.name][label.value] = True
                         # If mapping found add corresponding labels
                         try:
-                            for label_tuple in self._label_mapping[label.name][label.value]:
+                            for label_tuple in scraper_config['_label_mapping'][label.name][label.value]:
                                 extra_label = metric.label.add()
                                 extra_label.name, extra_label.value = label_tuple
                         except KeyError:
                             pass
 
-    def process_metric(self, message, **kwargs):
+    def process_metric(self, message, scraper_config, metric_transformers=None):
```

Comment:
The way it worked perviously with the "magic methods" was that these were defined for all check instances. It seems this now adds the functionality for these "magic methods" to now be defined differently for each check instance. Just confirming that this is how the `metric_transformers` works?

Reply:
Yes, `metric_transformers` is a dictionary of mappings of `<metric_name>: <function that is called when the metric name is encountered>'. This dictionary can be passed in when the actual check calls the `process` function on a scraper configuration.